distributed/http/semaphore/prometheus.py

```diff
@@ -0,0 +1,143 @@
+            "Note: if a semaphore is closed while there are still leases active, this count will not equal "
+            "`semaphore_acquired_total` after execution.",
+            labels=["name"],
+        )
+        for semaphore_name, count in sem_ext.metrics["release_total"].items():
+            semaphore_release_total.add_metric([semaphore_name], count)
+        yield semaphore_release_total
+
+        semaphore_time_to_acquire_lease = HistogramMetricFamily(
+            "semaphore_time_to_acquire_lease",
+            "Time it took to acquire a lease per semaphore (note: this only includes time spent on scheduler side, "
+            "it does"
+            " not include time spent on communication).",
+            labels=["name"],
+        )
+        histogram_buckets = (
+            0.005,
+            0.01,
+            0.025,
+            0.05,
+            0.075,
+            0.1,
+            0.25,
+            0.5,
+            0.75,
+            1.0,
+            2.5,
+            5.0,
+            7.5,
+            10.0,
+            INF,
+        )
+        for semaphore_name, list_of_timedeltas in sem_ext.metrics[
+            "time_to_acquire_lease"
+        ].items():
+            # numpy uses a different logic for interpreting bin edges than prometheus, so we add a `0` as lower bound
+            sample_count_per_bucket, _ = np.histogram(
+                list_of_timedeltas, bins=(0, *histogram_buckets)
+            )
+            # prometheus assumes cumulative histograms
+            sample_count_per_bucket = sample_count_per_bucket.cumsum()
+            # Create pairs of bucket name and value
+            bucket_name_value_pairs = list(
+                zip(map(str, histogram_buckets), sample_count_per_bucket)
+            )
+            semaphore_time_to_acquire_lease.add_metric(
+                [semaphore_name],
+                buckets=bucket_name_value_pairs,
+                sum_value=sum(list_of_timedeltas),
+            )
+
+        yield semaphore_time_to_acquire_lease
+
+
+class PrometheusHandler(RequestHandler):
+    _collector = None
+
+    def __init__(self, *args, dask_server=None, **kwargs):
+        import prometheus_client
+
+        super(PrometheusHandler, self).__init__(
+            *args, dask_server=dask_server, **kwargs
+        )
+
+        if PrometheusHandler._collector:
+            # Especially during testing, multiple schedulers are started
+            # sequentially in the same python process
+            PrometheusHandler._collector.server = self.server
+            return
+
+        PrometheusHandler._collector = _PrometheusCollector(self.server)
+        prometheus_client.REGISTRY.register(PrometheusHandler._collector)
+
+    def get(self):
+        import prometheus_client
+
+        self.write(prometheus_client.generate_latest())
+        self.set_header("Content-Type", "text/plain; version=0.0.4")
+
+
+routes = [("/metrics", PrometheusHandler, {})]
```

Comment:
I think that this route is already taken by `distributed.http.scheduler.prometheus`.  Probably we should do some runtime checks to make sure that there aren't collisions like this.

Reply:
It's also in https://github.com/dask/distributed/blob/c8ee44bebe40e4db4495be1e2e0340fafbb41a37/distributed/http/worker/prometheus.py#L97. It seems to work fine when accessing the `/metrics`endpoint (I see the metrics from both files)

Reply:
Those are never included in the same server.  The scheduler file is for the scheduler and the worker file is for the worker.   But presumably you want to include this route in one or the other of these two, and so there would be some conflict there.  When Prometheus comes by it will only be able to access one of these Handlers.

More generally though I like the idea of making our Prometheus metrics extensible.  I think that this particular approach isn't exactly the right way, but I would very much like to encourage you to think about how to resolve this problem.

Reply:
> I would very much like to encourage you to think about how to resolve this problem

Sure, I'll try to come up with something